content/blog/eng/functional-testing-symfony-guzzle.md

```diff
@@ -0,0 +1,319 @@
+
+      if($response->getStatusCode()===200){
+            return new JsonResponse(
+                [
+                    'status'=>'ok',
+                    'message' => $response->getBody()
+                ],
+                 200
+            );
+        }
+
+        if($response->getStatusCode()===404){
+            return new JsonResponse(
+                [
+                   'status' => 'error',
+                   'message' => 'foo not found'
+                ],
+                 404
+            );
+        }
+
+        return new JsonResponse(
+            [
+                'status'=>'error',
+                'message' => 'something went wrong'
+            ],
+            500
+        );
+
+    }
+```
+
+MyApiClient is done this way: it has a constructor that initialize the http client 
+and a method who retrieves foo from the API.
+
+```php
+<?php
+
+namespace MyNamespace;
+use Psr\Http\Client\ClientInterface;
+
+class MyApiClient
+{
+    public function __construct(ClientInterface $client)
+    {
+        $this->client = $client;
+    }
+
+    public function getFoo()
+    {
+        return $this->get('https://api-url/foo');
+    }
+
+}
+```
+
+Symfony will automatically inject MyApiClient in our controller, but to make sure 
+it will inject ClientInterface in MyApiClient we have to modify services.yaml to 
+look like this:
+
+```yaml
+Psr\Http\Client\ClientInterface: '@psr18.client'
+
+  psr18.client:
+    class: GuzzleHttp\Client
+```
+
+This way we are telling to symfony inject Guzzle Http Client in MyApiClient
+
+# Road to testing
+Now we have to create our controller's functional test. It will expect 3 different 
+responses according to API responses.
+But first…
+We create a class that extends Guzzle in order to easily manage the mock responses.  
+It will look like this:
+
+```php
+<?php
+
+namespace Tests\Fake;
+
+use GuzzleHttp\Handler\MockHandler;
+use GuzzleHttp\HandlerStack;
+use GuzzleHttp\Client;
+
+/**
+ * Class ClientFake
+ */
+class ClientFake extends Client
+{
+    /** @var MockHandler */
+    protected $mockHandler;
+
+    /**
+     * ClientFake constructor.
+     */
+    public function __construct()
+    {
+        $this->mockHandler = new MockHandler();
+        $handler = HandlerStack::create($this->mockHandler);
+        parent::__construct(['handler' => $handler]);
+    }
+
+    /**
+     * @param $responses
+     */
+    public function appendResponse($responses): void
+    {
+        $this->mockHandler->append(...$responses);
+    }
+}
+```
+
+This is not enough yet. We have to tell symfony to use this client instead 
+GuzzleHttp\Client when we are running our test. So we open our services_test.yml 
+and we add following lines
+
+```yaml
+psr18.client:
+    public: true
+    class: Tests\Fake\ClientFake
+```
+
+so when we are running our tests symfony will inject in MyApiClient ClientFake 
+instead of real GuzzleClient. Now are we ready to write our test? Almost! 
```

Comment:
Start a new line with the question and the answer. 
Within the question, move "now" at the end, as follows:

>  Are we ready to write our test now?